### PR TITLE
#43 review model and controller

### PR DESCRIPTION
**Pull request by yliu298**
### What this PR does (required):
- create review model and controller
-  @route POST /review/:id
-  @Given parameters passed in, create a review.
-  @route GET /review/:id
-  @Given profile id, get all reviews

### Screenshots / Videos (required):
- NA

### Any information needed to test this feature (required):
- will test it when review route is made

### Any issues with the current functionality (optional):
- NA

### CODE REVIEW

File: server/controllers/review.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+const ObjectId = require("mongoose").Types.ObjectId;
+const asyncHandler = require("express-async-handler");
+const Profile = require("../models/Profile");
+const User = require("../models/User");
+const Review = require("../models/Review");
+
+// @route POST /review/:id
+// @Given parameters passed in, create a review.
+exports.createReview = asyncHandler(async (req, res, next) => {
+  const userId = req.body.id;
```
Comment (dproc96) [inline server/controllers/review.js]: You should be able to get the current user id with req.user.id when protecting a route

File: server/controllers/review.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      userId
+    ).populate("profile");
+    const sitterProfile = await Profile.findById(profileId);
+    const newReview = new Review({
+      ...reviewData,
+      user: {
+        firstName: firstName,
+        lastName: lastName,
+        profileImg: profileImg,
+      },
```
Comment (dproc96) [inline server/controllers/review.js]: This would make the data static, better to pass a reference to the user

File: server/controllers/review.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    });
+    res.status(200).jason({
```
Comment (dproc96) [inline server/controllers/review.js]:
```suggestion
    res.status(200).json({
```

File: server/controllers/review.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  if (!ObjectId(userId)) {
+    return res.status(400).send(Error("User id is invalid"));
+  }
+
+  try {
+    const { firstName, lastName, profileImg } = await User.findById(
+      userId
+    ).populate("profile");
+    const sitterProfile = await Profile.findById(profileId);
+    const newReview = new Review({
```
Comment (dproc96) [inline server/controllers/review.js:23]: This review isn't being saved at any point or linked to the sitter profile

File: server/controllers/review.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    res.status(500);
+    throw new Error(e.message);
+  }
+});
+
+// @route GET /review/:id
+// @Given profile id, get all reviews
+exports.getAllReviews = asyncHandler(async (req, res, next) => {
+  const profileId = req.params.id;
+  const userId = req.body.id;
```
Comment (dproc96) [inline server/controllers/review.js]: This variable isn't being used, you can delete it

File: server/controllers/review.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  if (!ObjectId(userId)) {
+    return res.status(400).send(Error("User id is invalid"));
+  }
+
+  try {
+    const sitterProfile = await Profile.findById(profileId).populate("review");
+    res.status(200).jason({
```
Comment (dproc96) [inline server/controllers/review.js]:
```suggestion
    res.status(200).json({
```

File: server/controllers/review.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      success: {
+        reviews: sitterProfile.review,
+      },
+    });
+  } catch (e) {
+    res.status(500);
+    throw new Error(e.message);
```
Comment (dproc96) [inline server/controllers/review.js:63]: You could handle this error here instead of throwing it

File: server/controllers/review.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  try {
+    const sitterProfile = await Profile.findById(profileId).populate("review");
+    res.status(200).jason({
+      success: {
+        reviews: sitterProfile.review,
+      },
+    });
+  } catch (e) {
+    res.status(500);
```
Comment (dproc96) [inline server/controllers/review.js:62]: This response is not being sent

File: server/models/Review.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    type: Number,
+    min: 1,
+    max: 5,
+    required: true,
+  },
+  message: String,
+  postDate: {
+    type: Date,
+    default: Date.now(),
+  },
```
Comment (dproc96) [inline server/models/Review.js]: Mongoose can do this by passing a timestamp parameter to the Schema declaration

**Review by dproc96 — CHANGES_REQUESTED** [review]
Had a few comments, please be sure to test your features before making a pull request as it will lead to quicker merging

File: server/controllers/review.js (hunk 1 above)
Comment (yliu298) [reply server/controllers/review.js]: changed to req.user.id

File: server/controllers/review.js (hunk 2 above)
Comment (yliu298) [reply server/controllers/review.js]: changed to pass a reference to the user

File: server/controllers/review.js (hunk 3 above)
Comment (yliu298) [reply server/controllers/review.js]: commit suggestion

File: server/controllers/review.js (hunk 4 above)
Comment (yliu298) [reply server/controllers/review.js:23]: updated to save new review and sitter profile, and push new review id to sitter profile

File: server/controllers/review.js (hunk 5 above)
Comment (yliu298) [reply server/controllers/review.js]: delete user id

File: server/controllers/review.js (hunk 6 above)
Comment (yliu298) [reply server/controllers/review.js]: commit suggestion

File: server/models/Review.js (hunk 9 above)
Comment (yliu298) [reply server/models/Review.js]: changed to timestamps: true

**Review by bonnieli — APPROVED** [review]
There's one more of Dan's comments left unaddressed. Plus please update this branch as there's some merge conflicts

File: server/controllers/review.js (hunk 8 above)
Comment (yliu298) [reply server/controllers/review.js:62]: updated to send response

**Comment by bonnieli** [comment]
Since PR has been approved, you can merge this at your own discretion

**Pull request merged**